libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -688,155 +693,182 @@ namespace pika::mpi::experimental {
-                    str<>("enabling polling"), "pool", get_pool_name(), detail::mode_string(mode)));
-            detail::register_polling(pika::resource::get_thread_pool(get_pool_name()));
+                debug(str<>("pool created"), "name", get_pool_name(), "mode flags", bin<8>(flags)));
+            //
+            register_pool(name);
+            return true;
         }
-    }
 
-    // -------------------------------------------------------------
-    void enable_optimizations(bool enable) { detail::mpi_data_.optimizations_ = enable; }
+        // -----------------------------------------------------------------
+        bool pool_exists() { return custom_pool_exists_; }
 
-    // -------------------------------------------------------------
-    // Return the "best" mpi thread mode to use for initialization
-    // if all requests are transferred to the mpi pool, then single threaded is ok
-    int get_preferred_thread_mode()
-    {
-        int required = MPI_THREAD_MULTIPLE;
-        if (detail::singlethreaded(get_completion_mode()) && detail::mpi_data_.optimizations_)
+        // -------------------------------------------------------------
+        void init()
         {
-            required = MPI_THREAD_SINGLE;
-            PIKA_DETAIL_DP(detail::mpi_debug<0>, debug(str<>("MPI_THREAD_SINGLE"), "overridden"));
+            int provided, required = get_preferred_thread_mode();
+            mpi::detail::environment::init(nullptr, nullptr, required, required, provided);
         }
-        return required;
-    }
+    }    // namespace detail
+
+    // -----------------------------------------------------------------
+    const std::string& get_pool_name() { return detail::polling_pool_name_; }
 
     // -----------------------------------------------------------------
     // initialize the pika::mpi background request handler
     // All ranks should call this function,
     // but only one thread per rank needs to do so
-    void init(bool init_mpi, bool init_errorhandler)
+    void start_polling(exception_mode errorhandler, std::string pool_name)
     {
         // don't allow polling code to run until init has completed
         std::lock_guard<detail::mutex_type> lk(detail::mpi_data_.polling_vector_mtx_);
+        PIKA_DETAIL_DP(detail::mpi_debug<1>, debug(str<>("start_polling"), detail::mpi_data_));
 
-        // --------------------------------------
-        // the user has asked us to call mpi_init
-        if (init_mpi)
+        if (pool_name.empty())
         {
-            int provided, required = get_preferred_thread_mode();
-            mpi::detail::environment::init(nullptr, nullptr, required, required, provided);
+            pool_name = detail::pool_exists() ? get_pool_name() :
+                                                resource::get_partitioner().get_default_pool_name();
         }
-        else if (!mpi::detail::environment::is_mpi_initialized())
+        detail::register_pool(pool_name);
```

Comment:
Essentially yes. If we let pika do everything, then create pool calls register and all is good, but if the user manually creates a pool and then starts polling, we need to make sure internal stuff is set and so an extra register is needed - I can make the user call register themselves in that case, but I've done my best to ensure that it is safe so that this is harmless as is